### PR TITLE
[no-release-notes] Docstrings for chunker

### PR DESCRIPTION
**Pull request by max-hoffman**
Logical refactor that makes it easier to add documentation for
chunker.AdvanceTo.

### CODE REVIEW

**Comment by max-hoffman** [comment]
benchmarks after
```
BenchmarkMapUpdate/benchmark_maps_10k/benchmark_new_format_writes
BenchmarkMapUpdate/benchmark_maps_10k/benchmark_new_format_writes-12         	   19617	     59245 ns/op	   19347 B/op	      54 allocs/op
BenchmarkMapUpdate/benchmark_maps_10k/benchmark_old_format_writes
BenchmarkMapUpdate/benchmark_maps_10k/benchmark_old_format_writes-12         	    1969	    637116 ns/op	  481815 B/op	    3173 allocs/op
BenchmarkMapUpdate/benchmark_maps_100k
BenchmarkMapUpdate/benchmark_maps_100k/benchmark_new_format_writes
BenchmarkMapUpdate/benchmark_maps_100k/benchmark_new_format_writes-12        	   19452	     62476 ns/op	   20893 B/op	      54 allocs/op
BenchmarkMapUpdate/benchmark_maps_100k/benchmark_old_format_writes
BenchmarkMapUpdate/benchmark_maps_100k/benchmark_old_format_writes-12        	    1320	   1015432 ns/op	  713799 B/op	    4907 allocs/op
BenchmarkMapUpdate/benchmark_maps_1M
BenchmarkMapUpdate/benchmark_maps_1M/benchmark_new_format_writes
BenchmarkMapUpdate/benchmark_maps_1M/benchmark_new_format_writes-12          	   13495	     89753 ns/op	   28831 B/op	      54 allocs/op
BenchmarkMapUpdate/benchmark_maps_1M/benchmark_old_format_writes
BenchmarkMapUpdate/benchmark_maps_1M/benchmark_old_format_writes-12          	    1263	   1028665 ns/op	  819051 B/op	    5157 allocs/op
BenchmarkProllySmallWrites
BenchmarkProllySmallWrites/benchmark_new_format_writes
BenchmarkProllySmallWrites/benchmark_new_format_writes-12                    	   21049	     56210 ns/op	   19158 B/op	      54 allocs/op
BenchmarkTypesSmallWrites
BenchmarkTypesSmallWrites/benchmark_old_format_writes
BenchmarkTypesSmallWrites/benchmark_old_format_writes-12                     	    1732	    594336 ns/op	  486973 B/op	    3219 allocs/op
BenchmarkProllyMediumWrites
BenchmarkProllyMediumWrites/benchmark_new_format_writes
BenchmarkProllyMediumWrites/benchmark_new_format_writes-12                   	   19257	     61866 ns/op	   20920 B/op	      54 allocs/op
BenchmarkTypesMediumWrites
BenchmarkTypesMediumWrites/benchmark_old_format_writes
BenchmarkTypesMediumWrites/benchmark_old_format_writes-12                    	    1269	    905638 ns/op	  711083 B/op	    4872 allocs/op
BenchmarkProllyLargeWrites
BenchmarkProllyLargeWrites/benchmark_new_format_writes
BenchmarkProllyLargeWrites/benchmark_new_format_writes-12                    	   13502	     87961 ns/op	   28918 B/op	      54 allocs/op
```
benchmarks before:
```
BenchmarkMapUpdate/benchmark_maps_10k
BenchmarkMapUpdate/benchmark_maps_10k/benchmark_new_format_writes
BenchmarkMapUpdate/benchmark_maps_10k/benchmark_new_format_writes-12         	   20390	     58255 ns/op	   19321 B/op	      54 allocs/op
BenchmarkMapUpdate/benchmark_maps_10k/benchmark_old_format_writes
BenchmarkMapUpdate/benchmark_maps_10k/benchmark_old_format_writes-12         	    1990	    613679 ns/op	  482894 B/op	    3184 allocs/op
BenchmarkMapUpdate/benchmark_maps_100k
BenchmarkMapUpdate/benchmark_maps_100k/benchmark_new_format_writes
BenchmarkMapUpdate/benchmark_maps_100k/benchmark_new_format_writes-12        	   19617	     62682 ns/op	   20897 B/op	      54 allocs/op
BenchmarkMapUpdate/benchmark_maps_100k/benchmark_old_format_writes
BenchmarkMapUpdate/benchmark_maps_100k/benchmark_old_format_writes-12        	    1310	   1005681 ns/op	  713477 B/op	    4872 allocs/op
BenchmarkMapUpdate/benchmark_maps_1M
BenchmarkMapUpdate/benchmark_maps_1M/benchmark_new_format_writes
BenchmarkMapUpdate/benchmark_maps_1M/benchmark_new_format_writes-12          	   13424	     88028 ns/op	   28905 B/op	      54 allocs/op
BenchmarkMapUpdate/benchmark_maps_1M/benchmark_old_format_writes
BenchmarkMapUpdate/benchmark_maps_1M/benchmark_old_format_writes-12          	    1167	    983159 ns/op	  812032 B/op	    5046 allocs/op
BenchmarkProllySmallWrites
BenchmarkProllySmallWrites/benchmark_new_format_writes
BenchmarkProllySmallWrites/benchmark_new_format_writes-12                    	   21463	     55591 ns/op	   19042 B/op	      54 allocs/op
BenchmarkTypesSmallWrites
BenchmarkTypesSmallWrites/benchmark_old_format_writes
BenchmarkTypesSmallWrites/benchmark_old_format_writes-12                     	    2000	    595175 ns/op	  482842 B/op	    3172 allocs/op
BenchmarkProllyMediumWrites
BenchmarkProllyMediumWrites/benchmark_new_format_writes
BenchmarkProllyMediumWrites/benchmark_new_format_writes-12                   	   19276	     61658 ns/op	   20923 B/op	      54 allocs/op
BenchmarkTypesMediumWrites
BenchmarkTypesMediumWrites/benchmark_old_format_writes
BenchmarkTypesMediumWrites/benchmark_old_format_writes-12                    	    1173	    965731 ns/op	  717874 B/op	    4952 allocs/op
BenchmarkProllyLargeWrites
BenchmarkProllyLargeWrites/benchmark_new_format_writes
BenchmarkProllyLargeWrites/benchmark_new_format_writes-12                    	   13621	     87070 ns/op	   28886 B/op	      54 allocs/op
```

File: go/store/prolly/tree/chunker.go [hunk 1]
```diff
@@ -169,73 +179,65 @@ func (tc *chunker[S]) AdvanceTo(ctx context.Context, next *Cursor) error {
...
+		// end of tree
+		tc.cur.copy(next)
+		return nil
+	}
 
-				// |tc.cur| is now inconsistent with its parent, Invalidate it.
-				tc.cur.invalidate()
-			}
+	if tc.cur.parent.Compare(next.parent) == 0 { // step (3)
+		// (rare) new chunkstore synchronized with old chunkstore at the
```
Comment (andrew-wm-arthur) [inline go/store/prolly/tree/chunker.go:209]:
```suggestion
		// (rare) new tree synchronized with old tree at the
```

File: go/store/prolly/tree/mutator.go [hunk 2]
```diff
@@ -26,6 +26,32 @@ type MutationIter interface {
...
+//   where the edit will be applied, and move the tracking cursor to
+//   that index.
+// - Advance the chunker and the second cursor to the new edit point.
+//   Refer to the chunker.AdvanceTo docstring for details.
+// - Add the edit to the chunker. This applies the edit to the in-progress
+//   NodeStore. The new NodeStore may expand or shrink relative to the
+//   old tree, but these details are internal to the chunker.
+//  - Repeat for every edit.
+//
+//  - Finalize the chunker and resolve the new NodeStore root Node.
```
Comment (andrew-wm-arthur) [inline go/store/prolly/tree/mutator.go:54]:
```suggestion
// ApplyMutations applies a sorted series of edits to a NodeStore,
// returning the new root Node.
//
// The algorithm is structured as follows:
//
// - Create a new chunker, the main interface for building a new
//   tree.
// - Create two cursors into the previous tree. Both cursors
//   track key indexes in the old keyspace. The first tracks where
//   a new edit will be applied relative to the old keyspace.
//   The second indicates the most recent edit in the new tree
//   relative to the old keyspace. The second cursor is embedded in
//   the chunker, maintained by the chunker, and necessary precedes
//   the first.
//
// - For every edit, first identify the key index in the old keyspace
//   where the edit will be applied, and move the tracking cursor to
//   that index.
// - Advance the chunker and the second cursor to the new edit point.
//   Refer to the chunker.AdvanceTo docstring for details.
// - Add the edit to the chunker. This applies the edit to the in-progress
//   NodeStore. The new tree may expand or shrink relative to the
//   old tree, but these details are internal to the chunker.
//  - Repeat for every edit.
//
//  - Finalize the chunker and resolve the tree's new root Node.
```

**Review by andrew-wm-arthur — APPROVED** [review]
Looks Good, a few corrections

File: go/store/prolly/tree/node_cursor.go [hunk 3]
```diff
@@ -272,7 +272,7 @@ func (cur *Cursor) atNodeStart() bool {
 // points to the last node item
 func (cur *Cursor) atNodeEnd() bool {
 	lastKeyIdx := int(cur.nd.count - 1)
-	return cur.idx == lastKeyIdx
+	return cur.idx >= lastKeyIdx
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/store/prolly/tree/node_cursor.go:275]: yeah this is stale, reverting

**Pull request merged**